File: packages/daemon/src/cli/workspace-init.ts
import { spawn } from "node:child_process";
import fs from "node:fs";
import path from "node:path";
import type { IssueKey } from "@legion/contracts";
import { type ProvisionIssueWorkspaceDeps, provisionIssueWorkspace } from "@legion/workspace";
import { DEFAULT_SLOW_COMMAND_TIMEOUT_SECONDS } from "../daemon/config";
import { SESSIONS_SUBPATH } from "../daemon/k8s-manifests";
import { ISSUE_KEY_PATTERN } from "../daemon/legion-state";
import { type CommandRunner, defaultRunner } from "../state/fetch";
import { CliError } from "./errors";
import { parseRepo } from "./review-threads";
import { readSecretPointer } from "./secret-pointer";

export interface WorkspaceInitCommandDeps {
  env: NodeJS.ProcessEnv;
  run: ProvisionIssueWorkspaceDeps["run"];
  installGhShim(root: string): Promise<string>;
  /** `fs.promises.mkdir(path, { recursive: true })`: idempotent, parents included. */
  mkdir(path: string): Promise<unknown>;
  /** `fs.existsSync`: whether the recorded session file is on the volume. */
  exists(path: string): boolean;
  log(line: string): void;
}

/** The command runner both pod-side CLI entry points hand `@legion/workspace`: `legion
 * workspace-init` for provisioning and `legion worker-shim` for the working-copy adoption. Every
 * command runs with the process's own environment (`env`: the pod's PATH, HOME, the image's tool
 * locations) beneath the env the caller supplies — `createProvisioningCredential`'s for the clone
 * and the fetch (`GIT_ASKPASS`, `GIT_TERMINAL_PROMPT`, `LEGION_PROVISIONING_TOKEN`, and the
 * `GIT_CONFIG_COUNT`/`GIT_CONFIG_KEY_n`/`GIT_CONFIG_VALUE_n` pairs that reset the clone's
 * credential-helper chain for those commands, LEGION-178); `JJ_USER`/`JJ_EMAIL` for the adoption —
 * the caller's winning. `defaultRunner` spawns with exactly the env it is given, and the caller's
 * env alone has no PATH: on the daemon host `createDaemonRunner` merges the resolved tool
 * environment the same way; a pod has no such wrapper, so this is where `jj` and `git` become
 * resolvable. */
export function processEnvRunner(
  env: NodeJS.ProcessEnv,
  runner: CommandRunner = defaultRunner
): ProvisionIssueWorkspaceDeps["run"] {
  return async (cmd, opts) => {
    const result = await runner(cmd, { ...opts, env: { ...env, ...opts?.env } });

    return { ...result, stderr: result.stderr ?? "" };
  };
}

/** `LEGION_WORKSPACE_INIT_LOCK_WAIT_SECONDS`: how long a `workspace-init` waits for another
 * pod's provisioning of the same repository. The daemon sets it on every pod it spawns
 * (`KubernetesRuntime.workspaceInitLockWaitSeconds`: its boot watchdog's registration deadline,
 * `worker_boot_timeout_seconds x worker_boot_registration_deadline_intervals`, plus one interval),
 * because the daemon is what bounds an initialising pod -- its probe reports one alive, its
 * watchdog re-arms through a lock wait for exactly that deadline, then retires the pod and spawns
 * the next generation. Sizing the wait from the same numbers means the init container never gives
 * up on a wait the daemon would still tolerate, whatever the deployment configures. The default
 * below is only for an invocation nobody is watching (a manual run without the variable): three
 * `DEFAULT_SLOW_COMMAND_TIMEOUT_SECONDS` budgets, covering a live holder's two network commands
 * (`jj git clone`, `jj git fetch`) at full budget plus its local ones. A dead holder releases the
 * instant its process is gone (see `withWorkspaceInitLock`), so the wait only ever guards a hung
 * one, and turns an endless wait into a failure that names the lock. */
const WORKSPACE_INIT_LOCK_WAIT_ENV = "LEGION_WORKSPACE_INIT_LOCK_WAIT_SECONDS";
const DEFAULT_WORKSPACE_INIT_LOCK_WAIT_SECONDS = 3 * DEFAULT_SLOW_COMMAND_TIMEOUT_SECONDS;

function workspaceInitLockWaitSeconds(env: NodeJS.ProcessEnv): number {
  const value = env[WORKSPACE_INIT_LOCK_WAIT_ENV];
  if (value === undefined) return DEFAULT_WORKSPACE_INIT_LOCK_WAIT_SECONDS;
  if (!/^[1-9][0-9]*$/.test(value)) {
    throw new CliError(
      `${WORKSPACE_INIT_LOCK_WAIT_ENV} must be a positive whole number of seconds (got ${JSON.stringify(value)})`
    );
  }
  return Number(value);
}

/** Serializes shared-clone provisioning across the tree volume's init containers: every issue's
 * pod provisions against one clone under `<root>/repos/github.com/<owner>/<repo>`, and two pods
 * admitted together would otherwise run `jj git clone`/`fetch` and the `git config` writes against
 * it at once (git's `.git/config` lock refuses the second writer). The lock is `flock(2)` — an
 * exclusive advisory lock on `<clone>.lock` on the volume itself, the only state two pods share —
 * taken by util-linux `flock` (in the image's debian base) and held by a child shell for exactly
 * this process's lifetime: the shell blocks on its stdin, so the lock releases when this process
 * releases it, or when this process dies and the pipe closes. No lease, no mtime, no takeover: a
 * holder that is alive holds, however long its provisioning takes; one that is dead holds
 * nothing. A refused non-blocking attempt logs one line, so a pod stuck behind another's
 * provisioning says so in its init-container log, then waits, bounded. */
async function withWorkspaceInitLock<T>(
  root: string,
  repo: string,
  waitSeconds: number,
  log: (line: string) => void,
  operation: () => Promise<T>
): Promise<T> {
  const lockPath = path.join(root, "repos", "github.com", `${repo}.lock`);
  await fs.promises.mkdir(path.dirname(lockPath), { recursive: true });
  let release = await holdFlock(lockPath, ["--nonblock"]);
  if (release === undefined) {
    log(`workspace-init: waiting for ${lockPath} (another pod is provisioning ${repo})`);
    release = await holdFlock(lockPath, ["--timeout", String(waitSeconds)]);
    if (release === undefined) {
      throw new CliError(
        `Timed out after ${waitSeconds} s waiting for workspace-init lock ${lockPath}`
      );
    }
  }
  try {
    return await operation();
  } finally {
    await release();
  }
}

/** Runs `flock <options> <lockPath> sh -c 'echo && read _'`: once flock holds the file's exclusive
 * lock the shell prints one line and blocks on its stdin, holding the lock (the locked descriptor
 * is the shell's) until that stdin closes. Resolves the releaser — closes the stdin, awaits the
 * exit — when the line arrives; `undefined` when flock exits 1 instead, its one conflict code
 * (`--nonblock` refused, or `--timeout` elapsed). Any other exit is an error naming it; flock's
 * own stderr is inherited. */
async function holdFlock(
  lockPath: string,
  options: string[]
): Promise<(() => Promise<void>) | undefined> {
  const child = spawn("flock", [...options, lockPath, "sh", "-c", "echo && read _"], {
    stdio: ["pipe", "pipe", "inherit"],
  });
  const acquired = Promise.withResolvers<number | undefined>();
  child.once("error", acquired.reject);
  child.stdout.once("data", () => acquired.resolve(undefined));
  child.once("exit", (code) => acquired.resolve(code ?? 1));
  const exitCode = await acquired.promise;
  if (exitCode === undefined) {
    return async () => {
      const exited = Promise.withResolvers<void>();
      child.once("exit", () => exited.resolve());
      child.stdin.end();
      await exited.promise;
    };
  }
  if (exitCode === 1) return undefined;
  throw new CliError(`flock ${options.join(" ")} ${lockPath} exited ${exitCode}`);
}

/** `legion workspace-init` (the Kubernetes runtime's init container): prepares an issue's jj
 * working copy on the tree's persistent volume before the main container's `worker-shim` starts —
 * the same provisioning path `TmuxRuntime` calls in-process (`provisionIssueWorkspace`), the
 * daemon's own `gh` shim (`installWorkerGhShim`), and the volume's fixed directories the main
 * container's mounts expect (`sessions` for its `subPath: sessions` mount, `gh` for
 * `GH_CONFIG_DIR`). Both flags are validated, and the provisioning token file is read, before
 * anything is installed or run. */
export async function cmdWorkspaceInit(
  flags: { issue: string; repo: string; root: string; credentialHelper: string },
  deps: WorkspaceInitCommandDeps
): Promise<void> {
  if (!ISSUE_KEY_PATTERN.test(flags.issue)) {
    throw new CliError(
      `--issue must be a Dispatch issue key like LEGION-1 (got ${JSON.stringify(flags.issue)})`
    );
  }
  parseRepo(flags.repo);
  const tokenFile = deps.env.LEGION_PROVISION_TOKEN_FILE;
  if (tokenFile === undefined) {
    throw new CliError("LEGION_PROVISION_TOKEN_FILE is not set");
  }
  const token = readSecretPointer("LEGION_PROVISION_TOKEN_FILE", tokenFile);
  const lockWaitSeconds = workspaceInitLockWaitSeconds(deps.env);
  await deps.installGhShim(flags.root);
  await deps.mkdir(path.join(flags.root, SESSIONS_SUBPATH));
  await deps.mkdir(path.join(flags.root, "gh"));
  // The same-agent invariant the tmux runtime enforces with a `stat` on the daemon host: a
  // recorded session file that is gone from the volume is a launch failure (this container exits
  // non-zero, the pod goes Failed, the daemon counts it), never a silent fresh agent -- which is
  // exactly what the pinned OMP does with a missing `--resume` path (exit 0, `ready`, new agent).
  // Checked before the repository lock: the volume is mounted either way, and a doomed pod must
  // not hold the shared clone's lock while it fails.
  const resumeSessionFile = deps.env.LEGION_RESUME_SESSION_FILE;
  if (resumeSessionFile !== undefined && !deps.exists(resumeSessionFile)) {
    throw new CliError(
      `Refusing to start ${flags.issue} fresh: recorded OMP session file is missing from the tree volume: ${resumeSessionFile}`
    );
  }
  const spec = await withWorkspaceInitLock(flags.root, flags.repo, lockWaitSeconds, deps.log, () =>
    provisionIssueWorkspace(flags.issue as IssueKey, {
      repo: flags.repo as `${string}/${string}`,
      stateDir: flags.root,
      provisioningToken: async () => token,
      credentialHelper: flags.credentialHelper,
      commandTimeoutMs: DEFAULT_SLOW_COMMAND_TIMEOUT_SECONDS * 1000,
      run: deps.run,
    })
  );
  deps.log(`workspace-init: ${spec.workspaceDir} on ${spec.bookmark}`);
}

File: packages/daemon-go/cmd/legion/workspace_init.go
package main

import (
	"context"
	"encoding/json"
	"errors"
	"flag"
	"fmt"
	"io"
	"io/fs"
	"math"
	"os"
	"os/exec"
	"path/filepath"
	"regexp"
	"strconv"
	"strings"
	"syscall"
	"time"

	legionclaim "github.com/sjawhar/legion/daemon/internal/claim"
	"github.com/sjawhar/legion/daemon/internal/runtime/workerbin"
	"github.com/sjawhar/legion/daemon/internal/workspace"
)

const workspaceProvisionUsage = "legion workspace-init provision --issue <KEY> --repo <owner>/<repo> [--root /legion] --credential-helper <git helper> --feed <dir>"

const (
	// workspaceLostExitCode is the status that tells the runtime the tree volume itself was lost —
	// neither the shared clone nor the recorded OMP session is on it — rather than that one launch
	// failed (packages/daemon/src/daemon/runtime.ts:87).
	workspaceLostExitCode = 3
	// lockWaitEnv bounds how long this init container waits for another pod's provisioning of the
	// same repository. The daemon sets it on every pod from its own registration deadline, so the
	// wait never gives up on a pod the daemon would still tolerate.
	lockWaitEnv = "LEGION_WORKSPACE_INIT_LOCK_WAIT_SECONDS"
	// provisionTokenFileEnv points `workspace-init fetch` at the mounted provisioning token.
	provisionTokenFileEnv = "LEGION_PROVISION_TOKEN_FILE"
	// defaultLockWaitSeconds is for an invocation no daemon sized: three slow-command budgets, a
	// live holder's clone and fetch at full budget plus its local commands (workspace-init.ts:61).
	defaultLockWaitSeconds = 3 * int64(workspace.CommandTimeout/time.Second)
)

var lockWaitPattern = regexp.MustCompile(`^[1-9][0-9]*$`)

// volumeLostError is the one failure the runtime tells apart, by workspaceLostExitCode.
type volumeLostError string

func (e volumeLostError) Error() string { return string(e) }

// runWorkspaceInit is `legion workspace-init`, the Kubernetes runtime's two init containers, which
// prepare an issue's jj workspace on the tree's persistent volume before the main container's
// worker-shim starts (packages/daemon/src/cli/workspace-init.ts). `fetch` is the first: the one
// process of the pod that holds the provisioning token, in a container that mounts nothing a tree
// agent can write. `provision` is the second: all the tree volume's work, in a container the
// provisioning Secret is not mounted in. Each one's log lines go to stdout and its refusals and
// failures to stderr — together the init log the runtime quotes — with exit 1, or 3 for a lost
// volume.
func runWorkspaceInit(ctx context.Context, args []string, stdout, stderr io.Writer) int {
	if len(args) == 0 {
		fmt.Fprintf(stderr, "usage: %s\n       %s\n", workspaceFetchUsage, workspaceProvisionUsage)
		return 2
	}
	var run func() error
	switch args[0] {
	case "fetch":
		flags := newFlags("workspace-init fetch", stderr)
		repo := flags.String("repo", "", "repository as <owner>/<name> (required)")
		feed := flags.String("feed", "", "the pod's feed directory, the container's own (required)")
		if code, ok := parseWorkspaceInitFlags(flags, args[1:], workspaceFetchUsage, stderr); !ok {
			return code
		}
		run = func() error { return workspaceFetch(ctx, *repo, *feed, stdout) }
	case "provision":
		flags := newFlags("workspace-init provision", stderr)
		issue := flags.String("issue", "", "Dispatch issue key, e.g. LEGION-1 (required)")
		repo := flags.String("repo", "", "repository as <owner>/<name> (required)")
		root := flags.String("root", "/legion", "tree volume root directory")
		credentialHelper := flags.String("credential-helper", "", "git credential helper written into the shared clone's config (required)")
		feed := flags.String("feed", "", "the pod's feed directory, which `workspace-init fetch` filled (required)")
		if code, ok := parseWorkspaceInitFlags(flags, args[1:], workspaceProvisionUsage, stderr); !ok {
			return code
		}
		run = func() error { return workspaceInit(ctx, *issue, *repo, *root, *credentialHelper, *feed, stdout) }
	default:
		fmt.Fprintf(stderr, "legion workspace-init: unknown subcommand %q\n", args[0])
		return 2
	}
	err := run()
	if err == nil {
		return 0
	}
	fmt.Fprintf(stderr, "legion workspace-init %s: %v\n", args[0], err)
	if errors.As(err, new(volumeLostError)) {
		return workspaceLostExitCode
	}
	return 1
}

// parseWorkspaceInitFlags parses one subcommand's flags, refusing a positional argument; a false
// ok carries the exit code.
func parseWorkspaceInitFlags(flags *flag.FlagSet, args []string, usage string, stderr io.Writer) (code int, ok bool) {
	if err := flags.Parse(args); err != nil {
		return 2, false
	}
	if flags.NArg() != 0 {
		fmt.Fprintf(stderr, "legion %s: unexpected argument %q: %s\n", flags.Name(), flags.Arg(0), usage)
		return 2, false
	}
	return 0, true
}

// workspaceInit validates everything before it touches the volume, and refuses to run where the
// provisioning token is pointed at: this is the process that runs git and jj against what every
// agent of the tree can write. Then it installs the gh shim, creates the directories the main
// container mounts, holds a resume to the same agent, and provisions from the feed under the
// repository lock, which it holds until it returns.
func workspaceInit(ctx context.Context, issue, repo, root, credentialHelper, feed string, stdout io.Writer) error {
	if !legionclaim.IsIssueKey(issue) {
		return fmt.Errorf("--issue must be a Dispatch issue key like LEGION-1 (got %q)", issue)
	}
	if !filepath.IsAbs(root) {
		return fmt.Errorf("--root must be an absolute path (got %q)", root)
	}
	if credentialHelper == "" {
		return fmt.Errorf("--credential-helper is required: %s", workspaceProvisionUsage)
	}
	if !filepath.IsAbs(feed) {
		return fmt.Errorf("--feed must be an absolute path (got %q)", feed)
	}
	if _, set := os.LookupEnv(provisionTokenFileEnv); set {
		return errors.New(provisionTokenFileEnv + " is set: provisioning runs without the provisioning token, which `workspace-init fetch` alone holds")
	}
	lockWait, err := workspaceInitLockWait()
	if err != nil {
		return err
	}
	tools, err := provisioningTools("git", "jj")
	if err != nil {
		return err
	}
	located, err := workspace.Location(root, repo, issue)
	if err != nil {
		return err
	}
	cloneDir := located.Clone

	if err := workerbin.InstallGh(root); err != nil {
		return err
	}
	for _, dir := range []string{"sessions", "gh"} {
		if err := os.MkdirAll(filepath.Join(root, dir), 0o700); err != nil {
			return fmt.Errorf("create %s: %w", filepath.Join(root, dir), err)
		}
	}
	// A recorded session gone from the volume is a launch failure, never a silent fresh agent —
	// which is what OMP does with a missing --resume path. Checked before the repository lock: a
	// doomed pod must not hold the shared clone's lock while it fails.
	if session, set := os.LookupEnv("LEGION_RESUME_SESSION_FILE"); set {
		present, err := pathPresent(session)
		if err != nil {
			return fmt.Errorf("stat the recorded OMP session file %s: %w", session, err)
		}
		if !present {
			cloned, err := pathPresent(cloneDir)
			if err != nil {
				return fmt.Errorf("stat the shared clone %s: %w", cloneDir, err)
			}
			if !cloned {
				return volumeLostError(fmt.Sprintf("Tree volume for %s holds neither the clone (%s) nor the recorded OMP session file (%s): the volume was lost", issue, cloneDir, session))
			}
			return fmt.Errorf("Refusing to start %s fresh: recorded OMP session file is missing from the tree volume: %s", issue, session)
		}
	}

	release, err := lockRepository(ctx, cloneDir+".lock", repo, lockWait, stdout)
	if err != nil {
		return err
	}
	defer release()
	run := workspace.NewRunner(workspace.CommandTimeout, tools)
	provisioned, err := workspace.Provision(ctx, run, workspace.Request{
		StateDir: root, Repo: repo, Issue: issue, CredentialHelper: credentialHelper, Source: workspace.FromFeed(feed),
		Log: func(line string) { fmt.Fprintln(stdout, "workspace-init: "+line) },
	})
	if err != nil {
		return err
	}
	fmt.Fprintf(stdout, "workspace-init: %s on %s\n", provisioned.Dir, provisioned.Bookmark)
	if fromRef, set := os.LookupEnv("LEGION_WORKSPACE_RECOVERED_FROM"); set {
		return writeRecoveryMarker(ctx, run, provisioned.Dir, fromRef)
	}
	return nil
}

// workspaceInitLockWait is LEGION_WORKSPACE_INIT_LOCK_WAIT_SECONDS: a positive whole number of
// seconds, 900 when unset.
func workspaceInitLockWait() (int64, error) {
	value, set := os.LookupEnv(lockWaitEnv)
	if !set {
		return defaultLockWaitSeconds, nil
	}
	seconds, err := strconv.ParseInt(value, 10, 64)
	if !lockWaitPattern.MatchString(value) || err != nil || seconds > int64(math.MaxInt64/time.Second) {
		return 0, fmt.Errorf("%s must be a positive whole number of seconds (got %q)", lockWaitEnv, value)
	}
	return seconds, nil
}

// provisioningTools resolves the named tools a provisioning step runs from PATH — in an init
// container the image's, with no worker-bin shim or operator rc ahead of them — where the
// TypeScript runner found them (workspace-init.ts:36-45).
func provisioningTools(names ...string) (map[string]string, error) {
	tools := map[string]string{}
	for _, tool := range names {
		path, err := exec.LookPath(tool)
		if err != nil {
			return nil, fmt.Errorf("%s is not on PATH: %w", tool, err)
		}
		tools[tool] = path
	}
	return tools, nil
}

// lockPollInterval is how often a waiting init container tries the repository lock again.
const lockPollInterval = 250 * time.Millisecond

// lockRepository serializes provisioning across a tree volume's init containers: every issue's
// pod provisions against one shared clone, and two pods admitted together would otherwise run the
// clone, the fetch, and the git config writes against it at once (git's config lock refuses the
// second writer). The lock is flock(2) on <clone>.lock, the only state two pods share, held on this
// process's own descriptor until release — or until the process dies, since the kernel drops the
// lock with its last descriptor and no child inherits it (Go opens every file close-on-exec). No
// lease, no mtime, no takeover: a live holder holds, however long it takes; a dead one holds
// nothing. Every attempt is non-blocking (flock(2) promises waiters no order, so polling gives up
// nothing); the first refused one logs one line, so a pod stuck behind another's provisioning says
// so in its init log, and the attempts continue every lockPollInterval, bounded by waitSeconds and
// by ctx (workspace-init.ts:74-139).
func lockRepository(ctx context.Context, lockPath, repo string, waitSeconds int64, log io.Writer) (release func(), err error) {
	if err := os.MkdirAll(filepath.Dir(lockPath), 0o700); err != nil {
		return nil, fmt.Errorf("create %s: %w", filepath.Dir(lockPath), err)
	}
	// Read-write: where flock(2) is emulated with POSIX locks (NFS), an exclusive lock needs a
	// descriptor open for writing.
	file, err := os.OpenFile(lockPath, os.O_RDWR|os.O_CREATE, 0o600)
	if err != nil {
		return nil, fmt.Errorf("open workspace-init lock %s: %w", lockPath, err)
	}
	release = func() { _ = file.Close() }
	deadline := time.Now().Add(time.Duration(waitSeconds) * time.Second)
	for attempt := 0; ; attempt++ {
		err := flock(int(file.Fd()), syscall.LOCK_EX|syscall.LOCK_NB)
		if err == nil {
			return release, nil
		}
		if !errors.Is(err, syscall.EWOULDBLOCK) {
			release()
			return nil, fmt.Errorf("flock workspace-init lock %s: %w", lockPath, err)
		}
		if attempt == 0 {
			fmt.Fprintf(log, "workspace-init: waiting for %s (another pod is provisioning %s)\n", lockPath, repo)
		}
		remaining := time.Until(deadline)
		if remaining <= 0 {
			release()
			return nil, fmt.Errorf("Timed out after %d s waiting for workspace-init lock %s", waitSeconds, lockPath)
		}
		select {
		case <-ctx.Done():
			release()
			return nil, fmt.Errorf("stopped waiting for workspace-init lock %s: %w", lockPath, context.Cause(ctx))
		case <-time.After(min(lockPollInterval, remaining)):
		}
	}
}

func flock(fd, how int) error {
	for {
		if err := syscall.Flock(fd, how); !errors.Is(err, syscall.EINTR) {
			return err
		}
	}
}

// writeRecoveryMarker is the command side of workspace recovery: a relaunch after a lost volume
// names the ref it recovers from, and the recreated workspace records it with the commit it was
// recreated at in .legion/workspace-recovered.json (workspace-init.ts:196-215). recoveredAt is an
// ISO instant in milliseconds, UTC, as JavaScript's toISOString writes it.
func writeRecoveryMarker(ctx context.Context, run workspace.Runner, dir, fromRef string) error {
	result, err := workspace.RunChecked(ctx, run, []string{"jj", "log", "-r", "@", "--no-graph", "-T", "commit_id", "--color=never"}, nil, dir)
	if err != nil {
		return err
	}
	body, err := json.Marshal(struct {
		RecoveredAt string `json:"recoveredAt"`
		FromRef     string `json:"fromRef"`
		SHA         string `json:"sha"`
		Reason      string `json:"reason"`
	}{time.Now().UTC().Format("2006-01-02T15:04:05.000Z"), fromRef, strings.TrimSpace(result.Stdout), "volume-missing"})
	if err != nil {
		return err
	}
	markerDir := filepath.Join(dir, ".legion")
	if err := os.MkdirAll(markerDir, 0o755); err != nil {
		return fmt.Errorf("create %s: %w", markerDir, err)
	}
	if err := os.WriteFile(filepath.Join(markerDir, "workspace-recovered.json"), body, 0o644); err != nil {
		return fmt.Errorf("write the recovery marker: %w", err)
	}
	return nil
}

func pathPresent(path string) (bool, error) {
	_, err := os.Stat(path)
	if err == nil {
		return true, nil
	}
	if errors.Is(err, fs.ErrNotExist) {
		return false, nil
	}
	return false, err
}
